app/src/main/java/com/nervousfish/nervousfish/activities/QRExchangeKeyActivity.java

```diff
@@ -0,0 +1,251 @@
+     */
+    public AlertDialog getLastDialog() {
+        return lastDialog;
+    }
+
+
+
+    /**
+     * Shows the QR Code in an alert dialog popup screen
+     *
+     * @param qrCode The QR code to be shown.
+     */
+    private void showQRCode(final Bitmap qrCode) {
+
+        final LayoutInflater li = LayoutInflater.from(this);
+        final View myView = li.inflate(R.layout.qrcode, null);
+
+        final ImageView imageView = (ImageView) myView.findViewById(R.id.QRCodeImage);
+        imageView.setImageBitmap(qrCode);
+        final AlertDialog.Builder builder = new AlertDialog.Builder(this)
+                .setView(imageView)
+                .setPositiveButton(getString(R.string.popup_done), new DialogInterface.OnClickListener() {
+                    /**
+                     * {@inheritDoc}
+                     */
+                    @Override
+                    public void onClick(final DialogInterface dialog, final int which) {
+                        dialog.dismiss();
+                    }
+                });
+
+        ((ViewGroup) imageView.getParent()).removeView(imageView);
+        lastDialog = builder.create();
+        lastDialog.show();
+    }
+
+    /**
+     * Gets the public key of the user
+     * @return the public key.
+     */
+    public IKey getPublicKey() {
+        return publicKey;
+    }
+
+    /**
+     * Sets the publicKey.
+     * @param publicKey The key to be set.
+     */
+    public void setPublicKey(final IKey publicKey) {
+        this.publicKey = publicKey;
+    }
+
+
+    private final class EditNameClickListener implements DialogInterface.OnClickListener {
+
+        private final EditText editName;
+        private final IKey key;
+
+        /**
+         * Constructor for editname click listener.
+         * @param editName The textinput from which we show and get the name from.
+         * @param key The key made from the QR code.
+         */
+        private EditNameClickListener(final EditText editName, final IKey key) {
+            this.editName = editName;
+            this.key = key;
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        @Override
+        public void onClick(final DialogInterface dialog, final int which) {
+            try {
+                final String name = editName.getText().toString();
+
+                final Contact contact = new Contact(name, key);
+                final IDatabase database = serviceLocator.getDatabase();
+                database.addContact(contact);
+                dialog.dismiss();
+                final Intent intent = new Intent(QRExchangeKeyActivity.this, ContactActivity.class);
+                intent.putExtra(ConstantKeywords.SERVICE_LOCATOR, serviceLocator);
+                intent.putExtra(ConstantKeywords.CONTACT, contact);
+                QRExchangeKeyActivity.this.startActivity(intent);
+            } catch (final IOException e) {
+                LOGGER.error(e.getMessage());
```

Comment:
You can better use LOGGER.error("oops, asdfasdfasdf went wrong", e), because then the log will contain both the stacktrace And the message automatically (and also a beautiful message that asdfasdfasdf went wrong)